source/actions3/AMActionRunner3.cpp

```diff
@@ -476,6 +372,150 @@ void AMActionRunner3::onImmediateActionStateChanged(int state, int previousState
+	currentAction_ = newAction;
+	if (currentAction_) {
+		connect(currentAction_, SIGNAL(stateChanged(int,int)), this, SLOT(onCurrentActionStateChanged(int,int)));
+		connect(currentAction_, SIGNAL(progressChanged(double,double)), this, SIGNAL(currentActionProgressChanged(double,double)));
+		connect(currentAction_, SIGNAL(statusTextChanged(QString)), this, SIGNAL(currentActionStatusTextChanged(QString)));
+		connect(currentAction_, SIGNAL(expectedDurationChanged(double)), this, SIGNAL(currentActionExpectedDurationChanged(double)));
+	}
+
+	emit currentActionChanged(currentAction_);
+}
+
+bool AMActionRunner3::validateAction(AMAction3 *newAction, const QString &okButtonText, const QString &cancelButtonText )
+{
+	AMAction3::ActionValidity actionValidity = newAction->isValid();
+	if(actionValidity != AMAction3::ActionCurrentlyValid){
+		QString title;
+		QString message;
+		QString information;
+
+		if(actionValidity == AMAction3::ActionNotCurrentlyValid){
+			title = "The Action You've Added May Not Be Valid";
+			information = "\n\nYou can ignore this warning (maybe something will happen between now and the time the action is run to make it valid) or you can cancel adding the action.\nConsidering the action is about to run right now, this seems very unlikely.";
+		}
+		else if(actionValidity == AMAction3::ActionNeverValid){
+			title = "The Action You've Added Is Not Valid";
+			information = "\n\nYou cannot ignore this warning, this type of action is not supported.";
+		} else {
+			title = "The Action You've Added Is Valid";
+			information = "\n\nYou won't see this message.";
+		}
+
+		message = "The '" % newAction->info()->typeDescription() % "' action you're about to run reports:\n" % newAction->notValidWarning();
+
+		bool confirmToAddAction = AMErrorMon::showMessageBox(title, message, information, okButtonText, cancelButtonText, false); //default cancel
+		if (!confirmToAddAction) { // cancel butttton is clicked
+			cancelCurrentAction();
+			return false;
+		}
+	}
+
+	return true;
+}
+
+void AMActionRunner3::onCurrentActionStarting()
+{
+	if(loggingDatabase_)
+		currentAction_->setIsLoggingFinished(false);
+
+	AMListAction3* listAction = qobject_cast<AMListAction3*>(currentAction_);
+
+	if(listAction){
+
+		int parentLogId = -1;
+		AMListAction3* parentAction = qobject_cast<AMListAction3*>(listAction->parentAction());
+
+		if(parentAction)
+			parentLogId = parentAction->logActionId();
+
+		AMActionHistoryModel3 *historyModel = AMAppControllerSupport::actionHistoryModelFromDatabaseName(loggingDatabase()->connectionName());
+		if(!historyModel || !historyModel->logUncompletedAction(currentAction_, loggingDatabase_, parentLogId)){
+			AMErrorMon::report(AMErrorReport(this, AMErrorReport::Alert, -200, "There was a problem logging the uncompleted action to your database.  Please report this problem to the Acquaman developers."));
+		}
+	}
+
+	if (isScanAction())
+		emit scanActionCreated((AMScanAction *)currentAction());
```

Comment:
the function `isScanAction()` already did a `qobject_cast` check. It is safe to direct cast here.